src/python/pants/backend/terraform/tailor.py

```diff
@@ -21,28 +20,80 @@
 from pants.util.logging import LogLevel
 
 
+def longest_common_prefix(x, y):
+    """Find the longest common prefix between two sequences."""
+
+    i = j = 0
+    while i < len(x) and j < len(y):
+        if x[i] != y[j]:
+            break
+        i = i + 1
+        j = j + 1
+
+    return x[:i]
+
+
+def find_disjoint_longest_common_prefixes(raw_values):
```

Comment:
These should probably be moved into `core/goals/tailor.py` - I suspect we'll reuse this for other tailor implementations.

Also, type hints would be great. Even though this code is highly generic, I think it's fine to use `str` instead of adding a `TypeVar`, given that we only expect it to be used for paths. That will make the code easier to think about imo by making one part of the functionality more concrete. So something like `Sequence[str]`.

Reply:
I tried type hints originally with TypeVar but `mypy` had issues with them. Something something Sequence[Sequence[T]] something something.

Will just do them non-generic with `str`?

Reply:
> These should probably be moved into core/goals/tailor.py - I suspect we'll reuse this for other tailor implementations.

We can move them in a separate PR. For now, this is the only impl so going to leave here.

Reply:
Exactly, I recommend just `str`.

Reply:
technically `tuple[str]` but same difference